File: weather-library/src/main/java/org/rcgonzalezf/weather/openweather/model/WeatherInfoMapper.java
package org.rcgonzalezf.weather.openweather.model;

import android.support.annotation.NonNull;
import android.support.annotation.VisibleForTesting;
import java.util.ArrayList;
import java.util.List;
import org.rcgonzalezf.weather.common.models.WeatherInfo;

public class WeatherInfoMapper {

  private static final int ALL = -1;
  private List<ForecastData> mData;
  // I've found that most cases the forecast consists of 40 items
  private static final int INITIAL_SIZE = 40;

  public WeatherInfoMapper withData(@NonNull List<ForecastData> data) {
    mData = data;
    return this;
  }

  public @NonNull List<WeatherInfo> map() {
    return map(ALL);
  }

  @VisibleForTesting @NonNull List<WeatherInfo> map(int howMany) {
    int size = 10;
    if (howMany == ALL) {
      size = INITIAL_SIZE * mData.size();
    }
    int counter = 0;
    List<WeatherInfo> weatherInfoList = new ArrayList<>(size);
    for (ForecastData forecastData : mData) {

      for (WeatherData weather : forecastData.getWeatherList()) {
        WeatherInfo weatherInfo = new WeatherInfo();
        weatherInfo.setCityId(forecastData.getCity().getId());
        weatherInfo.setCityName(forecastData.getCity().getName());
        weatherInfo.setSpeed(weather.getSpeed());
        weatherInfo.setTemperature(weather.getTemp());
        weatherInfo.setHumidity(String.valueOf(weather.getHumidity()));
        weatherInfo.setDateTime(weather.getDateTime());
        weatherInfo.setWeatherId(weather.getWeatherId());
        weatherInfo.setCountry(forecastData.getCity().getCountry() == null ? "" : forecastData.getCity().getCountry());
        weatherInfo.setDeg(weather.getDeg());
        weatherInfo.setDescription(weather.getDescription());

        weatherInfoList.add(weatherInfo);
        ++counter;
        if (counter == howMany) {
          return weatherInfoList;
        }
      }
    }
    return weatherInfoList;
  }
}

File: weather-library/src/main/java/org/rcgonzalezf/weather/openweather/model/WeatherInfoMapper.kt
package org.rcgonzalezf.weather.openweather.model

import androidx.annotation.VisibleForTesting
import org.rcgonzalezf.weather.common.models.WeatherInfo
import java.util.ArrayList

class WeatherInfoMapper {

    companion object {
        private const val ALL = -1

        // I've found that most cases the forecast consists of 40 items
        private const val INITIAL_SIZE = 40
    }

    private var data: List<ForecastData>? = null
    fun withData(data: List<ForecastData>): WeatherInfoMapper {
        this.data = data
        return this
    }

    fun map(): List<WeatherInfo> {
        return map(ALL)
    }

    @VisibleForTesting
    fun map(howMany: Int): List<WeatherInfo> {
        var size = 10
        if (howMany == ALL) {
            size = INITIAL_SIZE * data!!.size
        }
        var counter = 0
        val weatherInfoList: MutableList<WeatherInfo> = ArrayList(size)
        for (forecastData in data!!) {
            for (weather in forecastData.getWeatherList()) {
                val weatherInfo = WeatherInfo()
                weatherInfo.setCityId(forecastData.city.id)
                weatherInfo.setCityName(forecastData.city.name)
                weatherInfo.setSpeed(weather.speed)
                weatherInfo.setTemperature(weather.temp)
                weatherInfo.setHumidity(weather.humidity.toString())
                weatherInfo.setDateTime(weather.dateTime)
                weatherInfo.setWeatherId(weather.weatherId)
                weatherInfo.setCountry(
                        if (forecastData.city.country == null) "" else forecastData.city.country)
                weatherInfo.setDeg(weather.deg)
                weatherInfo.setDescription(weather.description)
                weatherInfoList.add(weatherInfo)
                ++counter
                if (counter == howMany) {
                    return weatherInfoList
                }
            }
        }
        return weatherInfoList
    }
}
